File: ALL/Algo_Grading.py
def find_longest_subsequence(correct_order, student_answer):
    n = len(correct_order)
    m = len(student_answer)
    longest_subsequence = [0] * (m + 1)

    for i in range(1, n + 1):
        prev_longest = 0
        for j in range(1, m + 1):
            if correct_order[i - 1] == student_answer[j - 1]:
                longest_subsequence[j] = prev_longest + 1
            else:
                longest_subsequence[j] = max(longest_subsequence[j], longest_subsequence[j - 1])
            prev_longest = longest_subsequence[j]

    return longest_subsequence[m]

N, M = map(int, input().split())
correct_order = list(map(int, input().split()))
student_answers = [list(map(int, input().split())) for _ in range(M)]

scores = []
for answer in student_answers:
    score = find_longest_subsequence(correct_order, answer)
    scores.append(score)

for score in scores:
    print(score)

File: ALL/Algo_Grading.cpp
#include <bits/stdc++.h>
using namespace std;

using namespace std;

int find_longest_subsequence(int n, vector<int>& correct_order, vector<int>& answers) {
    vector<int> positions(n);
    for (int i = 0; i < n; i++) {
        positions[correct_order[i] - 1] = i;
    }
    
    vector<int> dp(n, 1);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < i; j++) {
            if (positions[answers[i] - 1] > positions[answers[j] - 1]) {
                dp[i] = max(dp[i], dp[j] + 1);
            }
        }
    }
    
    int max_length = 0;
    for (int i = 0; i < n; i++) {
        max_length = max(max_length, dp[i]);
    }
    
    return max_length;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    int N, M;
    cin >> N >> M;
    
    vector<int> correct_order(N);
    for (int i = 0; i < N; i++) {
        cin >> correct_order[i];
    }
    
    vector<int> scores;

    for (int i = 0; i < M; i++) {
        vector<int> answers(N);
        for (int j = 0; j < N; j++) {
            cin >> answers[j];
        }
        int score = find_longest_subsequence(N, correct_order, answers);
        scores.push_back(score);
    }

    for (int score : scores) {
        cout << score << endl;
    }

    return 0;
}
